Add what's new for 2.0.0-Beta4 (#4024)
JetBrains/kotlin-web-site: docs/topics/whatsnew-eap.md

* Update the Gradle version in your project to 8.3 when it becomes available.

### Smart cast improvements

The Kotlin compiler can automatically cast an object to a type in specific cases,
saving you the trouble of having to explicitly cast it yourself. This is called [smart casting](typecasts.md#smart-casts).
The Kotlin K2 compiler now performs smart casts in even more scenarios than before.

In Kotlin %kotlinEapVersion%, we've made improvements related to smart casts in the following areas:
* [Local variables and further scopes](#local-variables-and-further-scopes)
* [Type checks with logical `or` operator](#type-checks-with-logical-or-operator)
* [Inline functions](#inline-functions)
* [Properties with function types](#properties-with-function-types)
* [Exception handling](#exception-handling)
* [Increment and decrement operators](#increment-and-decrement-operators)

#### Local variables and further scopes

Previously, if a variable was evaluated as not `null` within an `if` condition, 

> the variable was smart cast and information about this variable was shared further within the scope of the `if` block. 

Reviewer: ```suggestion
the variable was smart cast, and information about this variable was shared further within the scope of the `if` block. 
```

However, if you declared the variable **outside** the `if` condition, no information about the variable was available 
within the `if` condition, so it couldn't be smart cast. This behavior was also seen with `when` expressions and `while` loops.

From Kotlin %kotlinEapVersion%, if you declare a variable before using it in your `if`, `when`, or `while` condition 
then any information collected by the compiler about the variable is accessible in the condition statement and its block for smart casting.
This can be useful when you want to do things like extract boolean conditions into variables.

> Then you can give the variable a meaningful name, which makes your code easier to read, and easily reuse the variable later in your code.

Reviewer: ```suggestion
Then, you can give the variable a meaningful name, which makes your code easier to read, and easily reuse the variable later in your code.
```

For example:

```kotlin
class Cat {
    fun purr() {
        println("Purr purr")
    }
}

fun petAnimal(animal: Any) {
    val isCat = animal is Cat
    if (isCat) {
        // In Kotlin %kotlinEapVersion%, the compiler can access
        // information about isCat, so it knows that
        // isCat was smart cast to type Cat.
        // Therefore, the purr() function is successfully called.
        // In Kotlin 1.9.20, the compiler doesn't know

>         // about the smart cast so calling the purr()

Reviewer: ```suggestion
        // about the smart cast, so calling the purr()
```

        // function triggers an error.
        animal.purr()
    }
}

fun main(){
    val kitty = Cat()
    petAnimal(kitty)
    // Purr purr
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0" id="kotlin-smart-casts-k2-local-variables" validate="false"}

#### Type checks with logical `or` operator

> In Kotlin %kotlinEapVersion%, if you combine type checks for objects with an `or` operator (`||`) then a smart cast 

Reviewer: ```suggestion
In Kotlin %kotlinEapVersion%, if you combine type checks for objects with an `or` operator (`||`), a smart cast 
```

is made to their closest common supertype. Before this change, a smart cast was always made to `Any` type. 
In this case, you still had to manually check the type of the object afterward before you could access any of its properties or call its functions.
For example:

```kotlin
interface Status {
    fun signal() {}
}

interface Ok : Status
interface Postponed : Status
interface Declined : Status

fun signalCheck(signalStatus: Any) {
    if (signalStatus is Postponed || signalStatus is Declined) {

>         // signalStatus is smart cast to common supertype: Status

Reviewer: ```suggestion
        // signalStatus is smart cast to a common supertype Status
```

        signalStatus.signal()
        // Prior to Kotlin %kotlinEapVersion%, signalStatus is smart cast 

>         // to type: Any so calling the signal() function, triggered an

Reviewer: ```suggestion
        // to type Any, so calling the signal() function triggered an
```

```

> The common supertype is an **approximation** of a union type. [Union types](https://en.wikipedia.org/wiki/Union_type)
> are not supported in Kotlin.
>
{type="note"}

#### Inline functions

In Kotlin %kotlinEapVersion%, the K2 compiler treats inline functions differently, 
allowing it to determine in combination with other compiler analyses whether it's safe to smart cast.

Specifically, inline functions are now treated as having an implicit 
[`callsInPlace`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.contracts/-contract-builder/calls-in-place.html) contract.
So any lambda functions passed to an inline function are called "in place". Since lambda functions are called in place, 
the compiler knows that a lambda function can't leak references to any variables contained within its function body. 
The compiler uses this knowledge along with other compiler analyses to decide if it's safe to smart cast any of the captured variables. 
For example:

```kotlin
interface Processor {
    fun process()
}

inline fun inlineAction(f: () -> Unit) = f()

fun nextProcessor(): Processor? = null

fun runProcessor(): Processor? {
    var processor: Processor? = null
    inlineAction {

>         // In Kotlin %kotlinEapVersion%, the compiler knows that processor 

Reviewer: ```suggestion
        // In Kotlin %kotlinEapVersion%, the compiler knows that the processor 
```

>         // is a local variable and inlineAction is an inline function so 

Reviewer: ```suggestion
        // is a local variable, and inlineAction() is an inline function, so 
```

>         // references to processor can't be leaked. Therefore, it's safe 

Reviewer: ```suggestion
        // references to the processor can't be leaked. Therefore, it's safe 
```

>         // to smart cast processor.

Reviewer: ```suggestion
        // to smart cast the processor.
```

>         // If processor isn't null, processor is smart cast

Reviewer: ```suggestion
        // If the processor isn't null, it is smart cast
```

        if (processor != null) {

>             // The compiler knows that processor isn't null so no safe call 

Reviewer: ```suggestion
            // The compiler knows that the processor isn't null, so no safe call 
```

            // is needed
            processor.process()

            // In Kotlin 1.9.20, you have to perform a safe call:
            // processor?.process()
        }

        processor = nextProcessor()
    }

    return processor
}
```

#### Properties with function types

In previous versions of Kotlin, it was a bug that class properties with a function type weren't smart cast.
We fixed this behavior in the K2 compiler in Kotlin %kotlinEapVersion%.
For example:

```kotlin
class Holder(val provider: (() -> Unit)?) {
    fun process() {

>         // In Kotlin %kotlinEapVersion%, if provider isn't null then

Reviewer: ```suggestion
        // In Kotlin %kotlinEapVersion%, if the provider isn't null, then
```

>         // provider is smart cast

Reviewer: ```suggestion
        // the provider is smart cast
```

        if (provider != null) {

>             // The compiler knows that provider isn't null

Reviewer: ```suggestion
            // The compiler knows that the provider isn't null
```

            provider()

>             // In 1.9.20 the compiler doesn't know that provider isn't 

Reviewer: ```suggestion
            // In 1.9.20, the compiler doesn't know that the provider isn't 
```

>             // null so it triggers an error:

Reviewer: ```suggestion
            // null, so it triggers an error:
```

            // Reference has a nullable type '(() -> Unit)?', use explicit '?.invoke()' to make a function-like call instead
        }
    }
}
```

This change also applies if you overload your `invoke` operator. For example:

```kotlin
interface Provider {
    operator fun invoke()
}

interface Processor : () -> String

class Holder(val provider: Provider?, val processor: Processor?) {
    fun process() {
        if (provider != null) {
            provider() 

>             // In 1.9.20 the compiler triggers an error: 

Reviewer: ```suggestion
            // In 1.9.20, the compiler triggers an error: 
```

            // Reference has a nullable type 'Provider?' use explicit '?.invoke()' to make a function-like call instead
        }
    }
}
```

#### Exception handling

In Kotlin %kotlinEapVersion%, we've made improvements to exception handling so that smart cast information can be passed
on to `catch` and `finally` blocks. This change makes your code safer as the compiler keeps track of whether 
your object has a nullable type or not.
For example:

```kotlin
//sampleStart
fun testString() {
    var stringInput: String? = null
    // stringInput is smart cast to String type
    stringInput = ""
    try {
        // The compiler knows that stringInput isn't null
        println(stringInput.length)
        // 0

        // The compiler rejects previous smart cast information for 

>         // stringInput. Now stringInput has String? type.

Reviewer: ```suggestion
        // stringInput. Now stringInput has the String? type.
```

        stringInput = null

        // Trigger an exception
        if (2 > 1) throw Exception()
        stringInput = ""
    } catch (exception: Exception) {
        // In Kotlin %kotlinEapVersion%, the compiler knows stringInput 

>         // can be null so stringInput stays nullable.

Reviewer: ```suggestion
        // can be null, so stringInput stays nullable.
```

        println(stringInput?.length)
        // null

        // In Kotlin 1.9.20, the compiler says that a safe call isn't
        // needed, but this is incorrect.
    }
}
//sampleEnd
fun main() {
    testString()
}
```
{kotlin-runnable="true" kotlin-min-compiler-version="2.0" id="kotlin-smart-casts-k2-exception-handling"}

#### Increment and decrement operators

Prior to Kotlin %kotlinEapVersion%, the compiler didn't understand that the type of an object can change after using 
an increment or decrement operator. As the compiler couldn't accurately track the object type, 
your code could lead to unresolved reference errors. In Kotlin %kotlinEapVersion%, this is fixed:

```kotlin
interface Rho {

>     operator fun inc(): sigma = TODO()

Reviewer 2: `Sigma`?

}

interface Sigma : Rho {
    fun sigma() = Unit
}

interface Tau {
    fun tau() = Unit
}

fun main(input: Rho) {
    var unknownObject: Rho = input

>     // Check if unknownObject inherits from interface tau

Reviewer: ```suggestion
    // Check if unknownObject inherits from the Tau interface
```

    if (unknownObject is Tau) {

>         // Uses the overloaded inc() operator from interface Rho

Reviewer: ```suggestion
        // Uses the overloaded inc() operator from interface Rho,
```

        // which smart casts the type of unknownObject to Sigma.
        ++unknownObject

        // In Kotlin %kotlinEapVersion%, the compiler knows unknownObject has type

>         // Sigma so the sigma() function is called successfully.

Reviewer: ```suggestion
        // Sigma, so the sigma() function is called successfully.
```

        unknownObject.sigma()

        // In Kotlin 1.9.20, the compiler thinks unknownObject has type

>         // Tau so calling the sigma() function throws an error.

Reviewer: ```suggestion
        // Tau, so calling the sigma() function throws an error.
```

        // In Kotlin %kotlinEapVersion%, the compiler knows unknownObject has type

>         // Sigma so calling the tau() function throws an error.

Reviewer: ```suggestion
        // Sigma, so calling the tau() function throws an error.
```

        unknownObject.tau()
        // Unresolved reference 'tau'

        // In Kotlin 1.9.20, the compiler mistakenly thinks that 

>         // unknownObject has type Tau so the tau() function is

Reviewer: ```suggestion
        // unknownObject has type Tau, so the tau() function is
```

help with type-checks, and make it easier to include Kotlin code in JavaScript.

The Kotlin/Wasm compiler collects any [top-level functions](wasm-js-interop.md#functions-with-the-jsexport-annotation) 
marked with `@JsExport` and automatically generates TypeScript definitions in a `.d.ts` file.

To generate TypeScript definitions, in your `build.gradle.kts` file in the wasmJs section, 
add the `generateTypeScriptDefinitions()` function:

```kotlin
kotlin {
    wasmJs {
        binaries.executable()
        browser {
        }
        generateTypeScriptDefinitions()
    }
}
```

## Kotlin/JS

This version brings the following changes:
* [Support for type-safe plain JavaScript objects](#support-for-type-safe-plain-javascript-objects)
* [Support for npm package manager](#support-for-npm-package-manager)
 
### Support for type-safe plain JavaScript objects

> The `js-plain-objects` plugin is [Experimental](components-stability.md#stability-levels-explained).
> It may be dropped or changed at any time. The `js-plain-objects` plugin **only** supports the K2 compiler.
>
{type="warning"}

> To make it easier to work with JavaScript APIs, in Kotlin 2.0.0-Beta4 we provide a new plugin: 

Reviewer: ```suggestion
To make it easier to work with JavaScript APIs, in Kotlin 2.0.0-Beta4, we provide a new plugin: 
```

[`js-plain-objects`](https://github.com/JetBrains/kotlin/tree/master/plugins/js-plain-objects),
that you can use to create type-safe plain JavaScript objects.
The plugin checks your code for any [external interfaces](wasm-js-interop.md#external-interfaces) 
that have a `@JsPlainObject` annotation and adds:

* an inline `invoke` operator function inside the companion object that you can use as a constructor.
* a `.copy()` function that you can use to create a copy of your object while adjusting some of its properties.

For example:

```kotlin
import kotlinx.js.JsPlainObject

@JsPlainObject
external interface User {
    var name: String
    val age: Int
    val email: String? 
}

fun main() {
    // Creates a JavaScript object
    val user = User(name = "Name", age = 10)
    // Copies the object and adds an email
    val copy = user.copy(age = 11, email = "some@user.com")

    println(JSON.stringify(user)) 
    // { "name": "Name", "age": 10 }
    println(JSON.stringify(copy)) 
    // { "name": "Name", "age": 11, "email": "some@user.com" }
}
```

Any JavaScript objects created with this approach are safer because instead of only seeing errors at runtime, 

> you can see them at compile time, or even highlighted by your IDE.

Reviewer: ```suggestion
you can see them at compile time or even highlighted by your IDE.
```

> Consider this example that uses a `fetch()` function to interact with a JavaScript API, using external interfaces 

Reviewer: ```suggestion
Consider this example that uses a `fetch()` function to interact with a JavaScript API using external interfaces 
```

to describe the shape of the JavaScript objects:

```kotlin
import kotlinx.js.JsPlainObject

@JsPlainObject
external interface FetchOptions {
    val body: String?
    val method: String
}

// A wrapper for Window.fetch
suspend fun fetch(url: String, options: FetchOptions? = null) = TODO("Add your custom behavior here")

> // A compile-time error is triggered as metod is not recognized

Reviewer: ```suggestion
// A compile-time error is triggered as the method is not recognized
```

Author: The use of `metod` is intentional but hard to show in code blocks without inline code formatting. I added some extra text to try and make it clear that `metod` is not recognized as `method`. The use of `metod` was undetected before but is now reported as an error with this change. I will mark other comments related to this as resolved.

> fetch("https://google.com", options = FetchOptions(metod = "POST")) 

Reviewer: ```suggestion
fetch("https://google.com", options = FetchOptions(method = "POST")) 
```

> // A compile-time error is triggered as method is required

Reviewer: ```suggestion
// A compile-time error is triggered as the method is required
```

fetch("https://google.com", options = FetchOptions(body = "SOME STRING")) 
```

In comparison, if you use the `js()` function instead to create your JavaScript objects,
errors are only found at runtime or aren't triggered at all:

```kotlin
suspend fun fetch(url: String, options: FetchOptions? = null) = TODO("Add your custom behavior here")

> // No error is triggered. As metod is not recognized, the wrong method 

Reviewer: ```suggestion
// No error is triggered. As the method is not recognized, the wrong method 
```

// (GET) is used.

> fetch("https://google.com", options = js("{ metod: 'POST' }")) 

Reviewer: ```suggestion
fetch("https://google.com", options = js("{ method: 'POST' }")) 
```

Author: This would break the example so I'm not including it.

// By default, the GET method is used. A runtime error is triggered as 
// body shouldn't be present.
fetch("https://google.com", options = js("{ body: 'SOME STRING' }")) 
// TypeError: Window.fetch: HEAD or GET Request cannot have a body
```

To use the `js-plain-objects` plugin, add the following to your `build.gradle.kts` file:

<tabs group="build-script">
<tab title="Kotlin" group-key="kotlin">

```kotlin
plugins {
    kotlin("plugin.js-plain-objects") version "%kotlinEapVersion%"
}
```

</tab>
<tab title="Groovy" group-key="groovy">

```groovy
plugins {
    id "org.jetbrains.kotlin.plugin.js-plain-objects" version "%kotlinEapVersion%"
}
```

</tab>
</tabs>

### Support for npm package manager

Previously, it was only possible for the Kotlin Multiplatform Gradle plugin to use [Yarn](https://yarnpkg.com/lang/en/) 
as a package manager to download and install npm dependencies. From Kotlin %kotlinEapVersion%, you can use [npm](https://www.npmjs.com/) 
as your package manager instead. Using npm as a package manager means that you have one less tool to manage during your setup. 

> For backwards compatibility, Yarn is still the default package manager. To use npm as your package manager, 

Reviewer: ```suggestion
For backward compatibility, Yarn is still the default package manager. To use npm as your package manager, 
```
